File: 21. Merge Two Sorted Lists.cpp
21. Merge Two Sorted Lists 
My Submissions QuestionEditorial Solution
Total Accepted: 120589 Total Submissions: 342161 Difficulty: Easy
Merge two sorted linked lists and return it as a new list. The new list should be made by splicing together the nodes of the first two lists.

Subscribe to see which companies asked this question



/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(l1 != NULL && l2 != NULL) {
            ListNode *p = l1;
            ListNode *q = l2;
            ListNode *t, *h;//t为新链表的连接指针，h为新链表的头指针
            if(p->val > q->val) {
                t = q;
                h = q;
                q = q->next;
            } else {
                t = p;
                h = p;
                p = p->next;
            }
            while(p != NULL && q != NULL) {
                if(p->val > q->val) {
                    t->next = q;
                    t = t->next;
                    q = q->next;
                } else {
                    t->next = p;
                    t = t->next;
                    p = p->next;
                }
                
            }
            while(p != NULL && q == NULL) {
                t->next = p;
                p = p->next;
                t = t->next;
            }
            while(p == NULL && q != NULL) {
                t->next = q;
                q = q->next;
                t = t->next;
            }
            while(p == NULL && q == NULL) {
                return h;
            }
        }
        if(l1 == NULL && l2 != NULL) {
            return l2;
        }
        if(l1 != NULL && l2 == NULL) {
            return l1;
        }
        if(l1 == NULL && l2 == NULL) {
            return NULL;
        }
    }
};

File: Swift/21. Merge Two Sorted Lists.swift
// 21. Merge Two Sorted Lists
// 24 ms, 60.98%
func mergeTwoLists(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
    var arr = [Int](), p1 = l1, p2 = l2, r = l1

    while let q1 = p1, let q2 = p2 {
        if q1.val < q2.val {
            arr.append(q1.val)
            p1 = q1.next
        } else {
            arr.append(q2.val)
            p2 = q2.next
        }
    }
    while let q1 = p1 {
        arr.append(q1.val)
        p1 = q1.next
        if p1 == nil {
            q1.next = l2
            r = l1
        }
    }
    while let q2 = p2 {
        arr.append(q2.val)
        p2 = q2.next
        if p2 == nil {
            q2.next = l1
            r = l2
        }
    }

    p1 = r
    arr.forEach {
        p1?.val = $0
        p1 = p1?.next
    }

    return r
}